enhancements/microshift/feature-csi-snapshots.md

```diff
@@ -0,0 +1,572 @@
+points to a real snapshot in the storage back end.
+
+For manually pre-provisioned snapshots, a cluster administrator creates a number of
+VolumeSnapshotContent CRDs. These carry the details of the real volume snapshot in the storage system.
+
+The VolumeSnapshotContent CRD is not namespaced and is for use by a cluster administrator.
+
+```yaml
+apiVersion: snapshot.storage.k8s.io/v1
+kind: VolumeSnapshotContent
+metadata:
+  name: new-snapshot-content-test
+  annotations:
+    - snapshot.storage.kubernetes.io/allow-volume-mode-change: "true"
+spec:
+  deletionPolicy: Delete
+  driver: hostpath.csi.k8s.io
+  source:
+    snapshotHandle: 7bdd0de3-aaeb-11e8-9aae-0242ac110002
+  sourceVolumeMode: Filesystem
+  volumeSnapshotRef:
+    name: new-snapshot-test
+    namespace: default
+```
+
+### Risks and Mitigations
+
+- Increased overhead.  The CSI snapshot controller and sidecar increase the on-disk footprint by a
+total of 360MB.  At idle, the components use a negligible amount of CPU (>1% CPU on a 2 core machine)
+and roughly 18Mb.
+
+### Drawbacks
+
+## Design Details
+
+### CSI Snapshotter Components
+
+The total additional cluster components are:
+
+1. VolumeSnapshot CRD
+2. VolumeSnapshotContent CRD
+3. VolumeSnapshotClass CRD
+
+### CSI-Snapshot Controller
+
+The CSI Snapshot Controller is maintained SIG-Storage upstream. Downstream OCP releases of this container 
+image will be used by MicroShift.  The controller is a standalone component responsible for watching 
+`volumeSnapshot`, `volumeSnapshotContent`, and `volumeSnapshotClass` objects.  When a `volumeSnapshot` is created, 
+the controller will trigger a snapshot operation by creating a `volumeSnapshotConent` object, which will be 
+detected by the _CSI external-snapshotter sidecar_.
+
+MicroShift will embed the controller's manifests and deploy them during startup.  They are: 
+
+1. CSI VolumeSnapshot Controller Deployment
+2. CSI VolumeSnapshot Controller Service Account
+3. CSI VolumeSnapshot Controller ClusterRole and ClusterRoleBinding
+4. CSI VolumeSnapshot Controller Role and RoleBinding
+
+### CSI-External
+
+The CSI Snapshot Validation Webhook is maintained SIG-Storage upstream. Downstream OCP releases of this container 
+image will be used by MicroShift. The container image is specified by LVMS and is included in the _topolvm-controller_ 
+deployment. It watches for `volumeSnapshotContent` and `volumeSnapshotClass` events.  When an event is detected, it 
+makes the appropriate CSI gRPC call to LVMS via a shared unix socket.
+
+### CSI-Validation-Webhook
+
+The CSI Snapshot Validation Webhook is maintained SIG-Storage upstream. Downstream OCP releases of this container 
+image will be used by MicroShift.  The webhook serves as gatekeeper to `volumeSnapshot` CREATE and UPDATE events. 
+For the specifics of validation, _see_ 
+[Kubernetes KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1900-volume-snapshot-validation-webhook#validating-scenarios).
```

Comment:
both these chapter describe CSI Snapshot Validation Webhook deployment. It's odd to deploy it as part of _topolvm-controller_ deployment, so probably only `### CSI-Validation-Webhook` should remain.

Reply:
That's my mistake.  I should have been more clear that it's meant to be deployed discretely.

Reply:
> Also, is the webhook deployed by CSI Snapshot Controller Operator

Yes.